### PR TITLE
Smart contracts: Origination of smart-contracts

### PR DESCRIPTION
**Pull request by Zett98**
## Depends
- [x] #506

## Problem
Currently we can't originate the contracts

## Solution
add simple code that allows originating the contract

### CODE REVIEW

File: src/lambda_vm/compiler.ml [hunk 1]
```diff
@@ -31,7 +31,13 @@ let compile_prim prim =
   | Fst -> P_fst
   | Snd -> P_snd
 
-module Vars = Map_with_cardinality.Make (String)
+module Vars = Map_with_cardinality.Make (struct
+  include String
+  let to_yojson t = `String t
+  let of_yojson = function
+    | `String t -> Ok t
+    | Yojson.Safe.(#t) -> Error "failure parsing Vars.t"
```
Comment (EduardoRFS) [inline src/lambda_vm/compiler.ml:39]:
```ocaml
type t = string [@@deriving yojson]
```

File: src/core/state.ml [hunk 2]
```diff
@@ -18,28 +27,74 @@ let apply_tezos_operation t tezos_operation =
           let destination = key_hash in
           Ledger.deposit destination amount ticket ledger
         | Originated _ -> failwith "not implemented" in
-      { ledger } in
+      { t with ledger } in
```
Comment (EduardoRFS) [inline src/core/state.ml:30]: Please no `with`

File: src/core/address.mli [hunk 3]
```diff
@@ -1,5 +1,6 @@
 open Crypto
 type t [@@deriving eq, ord, yojson]
+val is_implicit : t -> bool
```
Comment (EduardoRFS) [inline src/core/address.mli:3]: remove this function

File: src/core/state.ml [hunk 4]
```diff
@@ -18,28 +27,74 @@ let apply_tezos_operation t tezos_operation =
...
+        ~address ~contract in
+    Ok
+      ( {
+          contract_storage;
+          ledger =
+            Ledger.burn t.ledger ~sender ~ticket ~amount:Amount.(of_int to_burn);
+        },
+        None )
+  | None, Contract_origination _ ->
+    Error `Contract_originator_must_be_implicit_account
```
Comment (EduardoRFS) [inline src/core/state.ml:91]: Change sender type from Address.t to Key_hash.t

File: src/core/user_operation.ml [hunk 5]
```diff
@@ -1,12 +1,16 @@
...
 type initial_operation =
-  | Transaction    of {
+  | Transaction          of {
       destination : Key_hash.t;
       amount : Amount.t;
       ticket : Ticket_id.t;
     }
-  | Tezos_withdraw of {
+  | Contract_origination of {
+      to_originate : Smart_contracts.Origination_payload.t;
```
Comment (EduardoRFS) [inline src/core/user_operation.ml:10]: Origination_payload should be defined here. Also call it just `Contract`

File: src/core/user_operation.ml [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
+      ticket : Ticket_id.t;
```
Comment (EduardoRFS) [inline src/core/user_operation.ml:11]: What does this ticket mean? If it's the internal amount that the contract can carry, I would say to remove this as we don't have defined how contracts will interact. And it's definitely not in the scope of this PR.

File: src/core/ledger.ml [hunk 7]
```diff
@@ -75,6 +75,18 @@ let deposit destination amount ticket t =
            (destination_balance + amount);
     withdrawal_handles = t.withdrawal_handles;
   }
+
+(* TODO: proper semantics ???  *)
+let burn t ~sender ~ticket ~amount =
+  let open Amount in
+  let sender_balance = balance sender ticket t in
+  {
+    t with
```
Comment (EduardoRFS) [inline src/core/ledger.ml:84]: No `with`

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

**Review by d4hines — COMMENTED** [review]
I'm not really part of this side of the project but I was casually reading this and had a few comments.

File: src/core/state.ml (hunk 4 above)
Comment (Zett98) [reply src/core/state.ml]: changed

File: src/lambda_vm/compiler.ml (hunk 1 above)
Comment (Zett98) [reply src/lambda_vm/compiler.ml]: changed

File: src/core/state.ml (hunk 2 above)
Comment (Zett98) [reply src/core/state.ml:30]: changed

File: src/core/address.mli (hunk 3 above)
Comment (Zett98) [reply src/core/address.mli]: removed

File: src/core/ledger.ml (hunk 7 above)
Comment (Zett98) [reply src/core/ledger.ml]: changed

File: src/core/user_operation.ml (hunk 5 above)
Comment (Zett98) [reply src/core/user_operation.ml:10]: I'd keep it in smart_contracts.ml for simplicity now

File: src/core/ledger.ml [hunk 8]
```diff
@@ -75,6 +75,18 @@ let deposit destination amount ticket t =
            (destination_balance + amount);
     withdrawal_handles = t.withdrawal_handles;
   }
+
+(* TODO: proper semantics ???  *)
```
[the remark this answers is not in the document]
Comment (Zett98) [reply src/core/ledger.ml]: Removed as out of scope rn

File: src/core/user_operation.ml (hunk 6 above)
Comment (Zett98) [reply src/core/user_operation.ml]: removed as out of scope.

File: src/core/smart_contracts.ml [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    [@@deriving yojson]
+
+    let make ~code ~storage ~originator =
+      { code; storage; originated_by = originator }
+
+    let to_string t = to_yojson t |> Yojson.Safe.pretty_to_string
+    let originated_by t = t.originated_by
+  end
+
+  module Raw = struct
```
[the remark this answers is not in the document]
Comment (Zett98) [reply src/core/smart_contracts.ml:57]: i'll rename it later to a better name.

File: src/bin/deku_cli.ml [hunk 10]
```diff
@@ -162,6 +176,88 @@ let create_transaction node_folder sender_wallet_file received_address amount
...
+  let doc =
+    "Originates a contract. Contract origination will be communicated to all \
+     known validators to be included in the next block." in
+  Term.info "originate-contract" ~version:"%\226\128\140%VERSION%%" ~doc ~exits
+    ~man
+
+let originate_contract node_folder contract_json initial_storage
+    sender_wallet_file =
+  let open Networking in
+  let module SM = Core.Smart_contracts in
```
Comment (EduardoRFS) [inline src/bin/deku_cli.ml:190]: don't

File: src/core/state.ml (hunk 2 above)
Comment (EduardoRFS) [reply src/core/state.ml:26]: still here

File: src/core/contract_storage.mli [hunk 11]
```diff
@@ -0,0 +1,11 @@
+open Tezos
+open Smart_contracts
+type t [@@deriving yojson]
+
+val empty : t
+val equal : t -> t -> bool
+val originate_contract :
+  t -> address:Contract_hash.t -> contract:Contract.t -> t
```
Comment (EduardoRFS) [inline src/core/contract_storage.mli:8]: Deku contract should not be using Tezos.Contract_hash, it's not a Tezos.Contract_hash

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: src/lambda_vm/ident.ml [hunk 12]
```diff
@@ -1,7 +1,8 @@
 (* TODO: should I care about overflowing this? *)
-type t = int
+type t = int [@@deriving yojson]
 
 let compare = Int.compare
+let equal = Int.equal
```
Comment (EduardoRFS) [inline src/lambda_vm/ident.ml:5]:
```suggestion
type t = int [@@deriving yojson, ord, eq]
```

File: src/lambda_vm/ident.mli [hunk 13]
```diff
@@ -1,6 +1,8 @@
-type t
+type t [@@deriving yojson]
 
 val compare : t -> t -> int
+val equal : t -> t -> bool
```
Comment (EduardoRFS) [inline src/lambda_vm/ident.mli:4]:
```suggestion
type t [@@deriving yojson, ord, eq]
```

File: src/core/user_operation.mli [hunk 14]
```diff
@@ -1,19 +1,21 @@
...
+  | Contract_origination of Contract_vm.Origination_payload.t
+  | Tezos_withdraw       of {
       owner : Tezos.Address.t;
       amount : Amount.t;
       ticket : Ticket_id.t;
     }
 type t = private {
   hash : BLAKE2B.t;
-  sender : Address.t;
+  sender : Key_hash.t;
```
Comment (EduardoRFS) [inline src/core/user_operation.mli:17]: As we're moving from Address.t to Key_hash, this should be called source
```suggestion
  source: Key_hash.t;
```

File: src/core/user_operation.ml [hunk 15]
```diff
@@ -1,34 +1,37 @@
...
+  | Tezos_withdraw       of {
       owner : Tezos.Address.t;
       amount : Amount.t;
       ticket : Ticket_id.t;
     }
 [@@deriving yojson]
 type t = {
   hash : BLAKE2B.t;
-  sender : Address.t;
+  sender : Key_hash.t;
```
Comment (EduardoRFS) [inline src/core/user_operation.ml:19]:
```suggestion
  source : Key_hash.t;
```

File: src/core/contract_vm.ml [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+  module Compiler : sig
+    val compile :
+      Origination_payload.t -> gas:int -> (Contract.t, string) result
+  end
+end
+
+module Lambda : VM = struct
+  module Contract = struct
+    type code = Lambda_vm.Ir.script [@@deriving yojson, eq]
```
Comment (EduardoRFS) [inline src/core/contract_vm.ml:21]:
```suggestion
    (* TODO: terminology, likely change Ir.script *)
    type code = Lambda_vm.Ir.script [@@deriving yojson, eq]
```

File: src/core/state.ml [hunk 17]
```diff
@@ -18,28 +23,54 @@ let apply_tezos_operation t tezos_operation =
           let destination = key_hash in
           Ledger.deposit destination amount ticket ledger
         | Originated _ -> failwith "not implemented" in
-      { ledger } in
+      { ledger; contract_storage = t.contract_storage } in
```
Comment (EduardoRFS) [inline src/core/state.ml:26]:
```suggestion
      { ledger; contract_storage } in
```

File: src/core/contract_address.ml [hunk 18]
```diff
@@ -0,0 +1,11 @@
+open Tezos
+open Crypto
+
+type t = Contract_hash.t [@@deriving yojson, eq, ord]
+
+(* contract address is a hash of user_operation hash *)
+let of_user_operation_hash (hash : BLAKE2B.t) =
+  hash |> BLAKE2B.to_raw_string |> BLAKE2B_20.hash
+
+let to_string = Contract_hash.to_string
```
Comment (EduardoRFS) [inline src/core/contract_address.ml:10]: This is not okay, as we're adding a proper abstraction for Contract_address, this should have it's own prefix.

File: src/core/contract_address.ml [hunk 19]
```diff
@@ -0,0 +1,11 @@
+open Tezos
+open Crypto
+
+type t = Contract_hash.t [@@deriving yojson, eq, ord]
+
+(* contract address is a hash of user_operation hash *)
+let of_user_operation_hash (hash : BLAKE2B.t) =
+  hash |> BLAKE2B.to_raw_string |> BLAKE2B_20.hash
```
Comment (EduardoRFS) [inline src/core/contract_address.ml:8]:
```suggestion
  (* TODO: likely BLAKE2B should expose a rehash function *)
  hash |> BLAKE2B.to_raw_string |> BLAKE2B_20.hash
```

File: src/core/state.ml [hunk 20]
```diff
@@ -18,28 +23,54 @@ let apply_tezos_operation t tezos_operation =
...
+  let { source; initial_operation; hash } = user_operation in
+  match initial_operation with
+  | Transaction { destination; amount; ticket } ->
     let%ok ledger =
-      Ledger.transfer ~sender ~destination amount ticket t.ledger in
-    Ok ({ ledger }, None)
-  | None, Transaction _ -> Error `Transaction_sender_must_be_implicit_account
-  | Some sender, Tezos_withdraw { owner; amount; ticket } ->
+      Ledger.transfer ~sender:source ~destination amount ticket t.ledger in
+    Ok ({ contract_storage = t.contract_storage; ledger }, None)
```
Comment (EduardoRFS) [inline src/core/state.ml:38]:
```suggestion
    Ok ({ ledger; contract_storage }, None)
```

File: src/core/state.ml [hunk 21]
```diff
@@ -18,28 +23,54 @@ let apply_tezos_operation t tezos_operation =
...
+    let wrap_error t = t |> Result.map_error (fun x -> `Origination_error x) in
+    (* TODO: Burn on storage size change, need CTEZ *)
+    let%ok contract =
+      Contract_vm.Compiler.compile ~gas:initial_gas to_originate |> wrap_error
+    in
+    let address = Contract_address.of_user_operation_hash hash in
+    let contract_storage =
+      Contract_storage.originate_contract t.contract_storage ~address ~contract
+    in
+    Ok ({ contract_storage; ledger = t.ledger }, None)
```
Comment (EduardoRFS) [inline src/core/state.ml:66]:
```suggestion
    Ok ({ ledger; contract_storage }, None)
```

File: src/core/state.ml [hunk 22]
```diff
@@ -18,28 +23,54 @@ let apply_tezos_operation t tezos_operation =
...
+
 let apply_user_operation t user_operation =
   match apply_user_operation t user_operation with
   | Ok (t, receipt) -> (t, receipt)
+  | Error (`Origination_error _) -> (t, None)
   | Error
       ( `Not_enough_funds | `Transaction_sender_must_be_implicit_account
-      | `Withdraw_sender_must_be_implicit_account ) ->
+      | `Withdraw_sender_must_be_implicit_account
+      | `Contract_originator_must_be_implicit_account ) ->
```
Comment (EduardoRFS) [inline src/core/state.ml:75]:
```suggestion) ->
```

File: src/core/state.ml [hunk 23]
```diff
@@ -18,28 +23,54 @@ let apply_tezos_operation t tezos_operation =
...
-  | None, Tezos_withdraw _ -> Error `Withdraw_sender_must_be_implicit_account
+      Ledger.withdraw ~sender:source ~destination:owner amount ticket t.ledger
+    in
+    Ok
+      ( { ledger; contract_storage = t.contract_storage },
+        Some (Receipt_tezos_withdraw handle) )
+  | Contract_origination to_originate ->
+    (* @TODO: deduct gas from account and check *)
+    let balance = max_int |> Amount.of_int in
+    let origination_cost = 250 |> Amount.of_int in
```
Comment (EduardoRFS) [inline src/core/state.ml:49]:
```suggestion
    (* TODO: those constants should not be defined here *)
    let origination_cost = 250 |> Amount.of_int in
```

File: src/protocol/protocol_operation.ml [hunk 24]
```diff
@@ -40,20 +40,17 @@ module Core_user = struct
   let sign ~secret ~nonce ~block_height ~data =
     let open User_operation in
     let key = Key.of_secret secret in
-    match Address.to_key_hash data.sender with
-    | Some sender when Key_hash.matches_key key sender ->
+    let sender = data.source in
```
Comment (EduardoRFS) [inline src/protocol/protocol_operation.ml:43]:
```suggestion
    let source= data.source in
```

File: src/protocol/protocol_operation.ml [hunk 25]
```diff
@@ -40,20 +40,17 @@ module Core_user = struct
...
+    else
       failwith
         "Operations to be signed should come from a key_hash that matches the \
          secret"
   let verify ~hash ~key ~signature ~nonce ~block_height ~data =
-    let%ok sender =
-      match Address.to_key_hash data.User_operation.sender with
-      | Some sender -> Ok sender
-      | _ -> Error "Sender should be a key_hash" in
+    let sender = data.User_operation.source in
```
Comment (EduardoRFS) [inline src/protocol/protocol_operation.ml:53]:
```suggestion
    let source = data.User_operation.source in
```

**Review by anchpop — APPROVED** [review]

File: src/core/state.ml [hunk 26]
```diff
@@ -18,28 +23,50 @@ let apply_tezos_operation t tezos_operation =
...
     let%ok ledger, handle =
-      Ledger.withdraw ~sender ~destination:owner amount ticket t.ledger in
-    Ok ({ ledger }, Some (Receipt_tezos_withdraw handle))
-  | None, Tezos_withdraw _ -> Error `Withdraw_sender_must_be_implicit_account
+      Ledger.withdraw ~sender:source ~destination:owner amount ticket ledger
+    in
+    Ok ({ ledger; contract_storage }, Some (Receipt_tezos_withdraw handle))
+  | Contract_origination to_originate ->
+    (* @TODO: deduct gas from account and check *)
+    let balance = max_int |> Amount.of_int in
```
Comment (EduardoRFS) [inline src/core/state.ml:47]: Int.max_int

File: src/core/state.ml [hunk 27]
```diff
@@ -18,28 +23,50 @@ let apply_tezos_operation t tezos_operation =
...
+    Ok ({ contract_storage; ledger }, None)
+
 let apply_user_operation t user_operation =
   match apply_user_operation t user_operation with
   | Ok (t, receipt) -> (t, receipt)
-  | Error
-      ( `Not_enough_funds | `Transaction_sender_must_be_implicit_account
-      | `Withdraw_sender_must_be_implicit_account ) ->
-    (t, None)
+  | Error (`Origination_error _) -> (t, None)
```
Comment (EduardoRFS) [inline src/core/state.ml:71]:
```suggestion
  (* TODO: use this origination_error for something *)
  | Error (`Origination_error _) -> (t, None)
```

**Review by EduardoRFS — APPROVED** [review]
Please just test to ensure it works, then feel free to merge.

weird anime girl approves

**Pull request merged**